fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/FallbackCursor.java

```diff
@@ -0,0 +1,163 @@
+     */
+    public FallbackCursor(@Nonnull RecordCursor<T> inner, @Nonnull Supplier<RecordCursor<T>> fallbackCursorSupplier) {
+        this.inner = inner;
+        this.fallbackCursorSupplier = fallbackCursorSupplier;
+        this.executor = inner.getExecutor();
+    }
+
+    @Nonnull
+    @Override
+    public CompletableFuture<RecordCursorResult<T>> onNext() {
+        try {
+            if (nextResultFuture != null && nextResultFuture.isDone() && !nextResultFuture.join().hasNext()) {
+                // This is needed to ensure we return the same terminal value once the cursor is exhausted
+                return nextResultFuture;
+            }
+        } catch (Exception ignored) {
+            // This will happen if the future finished exceptionally - just keep going (client will get
+            // the exception when they observe the future).
+        }
+        // The first stage (handle) will calculate the result of the operation if successful, or replace the inner
+        // with the fallback cursor if failed, and store future to the result in nextResultFuture.
+        // The second stage (thenCompose) will return nextResultFuture once the first stage is done.
+        return inner.onNext().handle((result, throwable) -> {
+            if (throwable == null) {
+                nextResultFuture = CompletableFuture.completedFuture(result);
+                // Cannot fail after the first result was delivered
+                allowedToFail = false;
+            } else {
+                if (alreadyFailed) {
+                    nextResultFuture = CompletableFuture.failedFuture(wrapException("Fallback cursor failed, cannot fallback again", throwable));
+                } else if (!allowedToFail) {
+                    nextResultFuture = CompletableFuture.failedFuture(wrapException("Cannot fallback to alternate cursor since inner already produced a record", throwable));
+                } else {
+                    inner.close();
+                    inner = fallbackCursorSupplier.get();
+                    nextResultFuture = inner.onNext();
+                }
+                alreadyFailed = true;
+            }
+            return null; // return value is ignored by next stage
+        }).thenCompose(vignore -> nextResultFuture);
+    }
+
+    @Override
+    public void close() {
+        inner.close();
+    }
+
+    @Nonnull
+    @Override
+    public Executor getExecutor() {
+        return this.executor;
+    }
+
+    private RecordCursor<T> getInner() {
+        return inner;
+    }
+
+    @Override
+    public boolean accept(@Nonnull RecordCursorVisitor visitor) {
+        if (visitor.visitEnter(this)) {
+            getInner().accept(visitor);
+        }
+        return visitor.visitLeave(this);
+    }
+
+    private Throwable wrapException(final String msg, final Throwable ex) {
+        if ((ex.getCause() != null) && (ex.getCause() instanceof LoggableException)) {
```

Comment:
Should this do something similar of `ex instanceof LoggableException`? The error handling around futures is always a bit tricky, and it's a little ambiguous whether the exception gets wrapped or not

Reply:
The `handle` method hands in a wrapped exception (by a `CompletionException`) - from `CompletableFuture.encodeThrowable`.
However, there should be no issue in defensive coding here to check for the `LoggableException` and a wrapper. 
Done.